File: hsr.jl
hsr(s :: AbstractString, sp :: AbstractString) = ∘((f -> make(f.([s, sp])...)).([identity, (s) -> join(collect(s), '.')])...)

make(s :: AbstractString, sp :: AbstractString) = (x) -> hsr0(s, sp, x)

function hsr0(s :: AbstractString, sp :: AbstractString, orig :: AbstractString) 
    i = 1
    r = ""
    lo = lowercase(orig)
    ls = lowercase(s)
    while true 
        p = findnext(ls, lo, i) 
        if p === nothing
            break
        end
        if p.start > i
            r *= orig[i:p.start-1]         
        end
        i = p.stop + 1
        r *= caseawarenessconvert(orig[p], sp)
    end
    if i <= length(orig)
        r *= orig[i:end]
    end
    return r
end

function caseawarenessconvert(s :: AbstractString, sp :: AbstractString)
    fs = ((c) -> if islowercase(c)
                    lowercase
                elseif isuppercase(c)
                    uppercase
                else
                    identity
                end).(collect(s))
    lsp = collect(sp)
    l = length(sp) > 1 ? (n -> 1 + round(Int, (length(fs) - 1) * (n - 1) / (length(lsp) - 1))).(1:length(lsp)) : 1:1
    join(broadcast((f, c) -> f(c), fs[l], lsp))
end

function test()
    f = hsr("windows", "harmony")
    println(f("Windows 10 is the newest W.I.N.D.O.W.S version"))
    println(f("WindoWS 2000 is built on NT"))
    println(f("WINdows is funny"))
end

test()

File: hsr.erl
-module(hsr).

-export([hsr/2]).
-export([test/0]).

-import(lists, [flatten/1, split/2]).

hsr(S, SP) when is_list(S), is_list(SP) ->
    L = [fun (Str) -> hsr(F(S), F(SP), Str) end || F <- [fun (X) -> X end, fun (X) -> string:join([[C] || C <- X], ".") end]],
    fun (X) -> lists:foldl(fun (F, Acc) -> F(Acc) end, X, L) end.

hsr(S, SP, Orig) -> hsr(string:casefold(S), string:casefold(Orig), SP, Orig, []).

hsr(LS, LO, SP, Orig, Acc) ->
    case string:find(LO, LS) of
        nomatch -> flatten(lists:reverse([Orig | Acc]));
        Match ->
            {S0, Rem} = split(length(Orig) - length(Match), Orig),
            {S1, OrigRem} = split(length(LS), Rem),
            hsr(LS, Rem, SP, OrigRem, [caseawarenessconvert(S1, SP), S0 | Acc])
    end.

caseawarenessconvert(S, SP) ->
    Fs = [get_f(C) || C <- S],
    L = case length(SP) of
        1 -> [1];
        _ -> [1 + round((length(Fs) - 1) * N / (length(SP) - 1)) || N <- lists:seq(0, length(SP) - 1)]
        end,
    flatten(lists:zipwith(fun (N, C) -> F = lists:nth(N, Fs), F([C]) end, L, SP)).

get_f(C) ->
    S = [C],
    case {S =:= string:uppercase(S), S =:= string:lowercase(S)} of
        {true, false} -> fun string:uppercase/1;
        {false, true} -> fun string:lowercase/1;
        _  -> fun (X) -> X end
    end.

test() ->
    F = hsr("windows", "harmony"),
    io:format("~ts~n", [F("Windows 10 is the newest W.I.N.D.O.W.S version")]),
    io:format("~ts~n", [F("WindoWS 2000 is built on NT")]),
    io:format("~ts~n", [F("WINdows is funny")]).
